internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,163 @@ import (
 	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
 )
 
+
+
+// Runtime2 is the VM.
+// Review: the object that implements this outlives actor calls.
+type Runtime2 struct {
+	CurerntEpoch() Epoch
+	// Review: why is this taking an offset? verify size of offset
+	Randomness(epoch Epoch, offset uint64) Randomness
+	// Review: the spec still has a TODO on this
+	// Review: not sure why the original has a messagereceipt instead of an InvocOutput
+	Send(input InvocInput) InvocOutput
+	// Review: how does this work? What qualifies as an "error"?
+	// Review: how is this any different than the caller just ignoring a returned error?
+	// SendAllowingErrors(input msg.InvocInput) msg.MessageReceipt
+
+	// Review: we might want a global Storage() Storage here
+}
+
+// InvocationContext is passed to the actors on each method call.
+// Review: this type + Runtime2 is mostly what the spec defines as Runtime.
+type InvocationContext interface {
+	// Review: grouped the VM level things out of the invocation itself
+	Runtime() Runtime2
+	Caller() address.Address
+	// Review: (is in spec, dont think we need it) idem to Caller().Equal(addr) :/
+	// ValidateCallerIs(caller addr.Address)
+	// Review: not sure what all this pattern matching adds up to, the "singleton" pattern does an equality check
+	// ValidateCallerMatches(CallerPattern)
+	// Review: the spec handles state and state changes a bit differently
+	// Review: our Storage is a bit of a mix of AcquireState and the Ipld{Get-Put}
+	// AcquireState() ActorStateHandle
+
+	// Review: The spec suggest handling return construction by the runtime, interesting but not strictly necessary
+	// SuccessReturn()     msg.InvocOutput
+	// ValueReturn(Bytes)  msg.InvocOutput
+	// ErrorReturn(exitCode exitcode.ExitCode) msg.InvocOutput
+
+	// Review: how does this work? is the spec expecting us to Panic? or just terminate the message execution?
```

Comment:
The runtime will panic, causing the actor code to immediately abort (and thus the return value isn't needed), but the runtime will then catch the error and convert it into an ordinary failure exit code. (See `Abort`, `_throwErrorFull`, and `_catchRuntimeErrors` in `systems/filecoin_vm/runtime/runtime.go`.)

Reply:
Unless there is a need for it to really be a method dispatched on the VMContext I would take it out of the interface. Most languages will not want to resolve this as a method you call on an interface, since you want to take over control of the callers execution flow, and some languages might not want that to be done with exceptions, or recovering from panics.  

somewhere in the vm package/namespace/module:
```
// Abort this execution of the vm call stack. 
// The caller should receive error 777 `PanicError`. 
func abort(message string) 
```

actor code..
```go
func somemethod(r Runtime) Result {
// ...
abort("blahblah")
// ...
}
```

most likely it will be implemented as `return abort("blahblah")` in most places and have it build the correct `InvOutput`, with it being actual panic/exceptions on languages that enjoy catching them all.

Reply:
We can still implement it the way you suggest, @icorderi.

I concur we probably want to reserve `panic/recover` for "faults", i.e. nondeterministic failures like OOM, disk full.

Reply:
A few notes here:

- Using return values instead of exceptions/panics may be error-prone in languages that make error-propagation difficult. E.g., actor method A calls helper function B (not itself an actor method), and B returns some combination of values indicating failure, but A fails to check for this condition. If B instead panics, A is forced to terminate also. Exceptions/panics are less idiomatic Go, but I believe are much less error-prone in cases like this.

- If we had language support for robust return-value-based error handling (e.g. Rust's `try` macros and option types), then I would suggest to use that instead of exceptions. In general, I think the conclusion is that the desired functionality is clearly specified, but in each language, we should use the most secure (not necessarily the most idiomatic) way of realizing that functionality. In Go, this seems to be panic/recover.

- If we allow panic/recover for nondeterministic faults such as disk full, then we have to decide what to do in this case. Presumably this means the runtime shouldn't handle the panic (since it's not part of the VM/actor logic), and just crash the process as a normal panic would. I think we can achieve this functionality, while still using panic/recover for ordinary VM errors also, by using different exception types (see `_catchRuntimeErrors` in `runtime.go`).